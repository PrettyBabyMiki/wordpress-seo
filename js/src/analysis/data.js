import debounce from "lodash/debounce";

/**
 * Represents the data.
 */
class Data {
	/**
	 *
	 *
	 * @param {Object} wpData The Gutenberg data API.
	 * @param {Object} YoastSEOApp The app of YoastSEO.
	 * @returns {void}
	 */
	constructor( wpData, YoastSEOApp ) {
		this._wpData = wpData;
		this._yoastSEOApp = YoastSEOApp;
		this.data = {};
		this.subscriber = debounce( this.refreshYoastSEO, 500 );
	}

	/**
	 * Checks whether the current data and the Gutenberg data are the same.
	 *
	 * @param {Object} currentData The current data.
	 * @param {Object} gutenbergData The data from Gutenberg.
	 * @returns {boolean} Whether the current data and the gutenbergData is the same.
	 */
	isShallowEqual( currentData, gutenbergData ) {
		if ( Object.keys( currentData ).length !== Object.keys( gutenbergData ).length ) {
			return false;
		}

		for( let dataPoint in currentData ) {
			if ( currentData.hasOwnProperty( dataPoint ) ) {
				if( ! ( dataPoint in gutenbergData ) || currentData[ dataPoint ] !== gutenbergData[ dataPoint ] ) {
					return false;
				}
			}
		}
		return true;
	}

	/**
	 * Retrieves the Gutenberg data for the passed post attribute.
	 *
	 * @param {string} attribute The post attribute you'd like to retrieve.
	 * @returns {string} The post attribute .
	 */
	getPostAttribute( attribute ) {
		return this._wpData.select( "core/editor" ).getEditedPostAttribute( attribute );
	}

	/**
	 * Collects the content, title, slug and excerpt of a post from Gutenberg.
	 *
	 * @param {Function} retriever The data retriever function.
	 * @returns {{content: string, title: string, slug: string, excerpt: string}} The content, title, slug and excerpt.
	 */
	collectGutenbergData( retriever ) {
		return {
			content: retriever( "content" ),
			title: retriever( "title" ),
			slug: retriever( "slug" ),
			excerpt: retriever( "excerpt" ),
		};
	}

	/**
	 * Refreshes YoastSEO's app when the Gutenberg data is dirty.
	 *
	 * @returns {void}
	 */
	refreshYoastSEO() {
		let gutenbergData = this.collectGutenbergData( this.getPostAttribute );

		// Set isDirty to false if the current data and Gutenberg data are unequal.
		let isDirty = ! this.isShallowEqual( this.data, gutenbergData );

		if ( isDirty ) {
			this.data = gutenbergData;
			this._yoastSEOApp.refresh();
		}
	}

	/**
	 * Listens to the Gutenberg data.
	 *
	 * @returns {void}
	 */
	subscribeToGutenberg() {
		// Fill data object on page load.
		this.data = this.collectGutenbergData( this.getPostAttribute );
		this._wpData.subscribe(
			this.subscriber
		);
	}

	/**
	 * Returns the data and whether the data is dirty.
	 *
	 * @returns {Object} The data and whether the data is dirty.
	 */
	getData() {
		console.log( this.data );
		return this.data;
	}
}
module.exports = Data;
