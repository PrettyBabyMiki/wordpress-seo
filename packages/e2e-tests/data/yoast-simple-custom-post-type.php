<?php
/**
 * Plugin Name:     Yoast_E2E_Post_Type
 * Author:          Team Yoast
 * Text Domain:     yoast_e2e_custom_plugin
 * Version:         0.0.0
 */

function yoast_e2e_register_yoast_post_type() {

	/**
	 * Post Type: Yoast Simple Posts.
	 */

	$labels = [
		"name" => __( "Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"singular_name" => __( "Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"menu_name" => __( "My Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"all_items" => __( "All Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"add_new" => __( "Add new", "yoast_e2e_custom_plugin" ),
		"add_new_item" => __( "Add new Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"edit_item" => __( "Edit Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"new_item" => __( "New Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"view_item" => __( "View Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"view_items" => __( "View Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"search_items" => __( "Search Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"not_found" => __( "No Yoast Simple Posts found", "yoast_e2e_custom_plugin" ),
		"not_found_in_trash" => __( "No Yoast Simple Posts found in trash", "yoast_e2e_custom_plugin" ),
		"parent" => __( "Parent Yoast Simple Post:", "yoast_e2e_custom_plugin" ),
		"featured_image" => __( "Featured image for this Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"set_featured_image" => __( "Set featured image for this Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"remove_featured_image" => __( "Remove featured image for this Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"use_featured_image" => __( "Use as featured image for this Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"archives" => __( "Yoast Simple Post archives", "yoast_e2e_custom_plugin" ),
		"insert_into_item" => __( "Insert into Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"uploaded_to_this_item" => __( "Upload to this Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"filter_items_list" => __( "Filter Yoast Simple Posts list", "yoast_e2e_custom_plugin" ),
		"items_list_navigation" => __( "Yoast Simple Posts list navigation", "yoast_e2e_custom_plugin" ),
		"items_list" => __( "Yoast Simple Posts list", "yoast_e2e_custom_plugin" ),
		"attributes" => __( "Yoast Simple Posts attributes", "yoast_e2e_custom_plugin" ),
		"name_admin_bar" => __( "Yoast Simple Post", "yoast_e2e_custom_plugin" ),
		"item_published" => __( "Yoast Simple Post published", "yoast_e2e_custom_plugin" ),
		"item_published_privately" => __( "Yoast Simple Post published privately.", "yoast_e2e_custom_plugin" ),
		"item_reverted_to_draft" => __( "Yoast Simple Post reverted to draft.", "yoast_e2e_custom_plugin" ),
		"item_scheduled" => __( "Yoast Simple Post scheduled", "yoast_e2e_custom_plugin" ),
		"item_updated" => __( "Yoast Simple Post updated.", "yoast_e2e_custom_plugin" ),
		"parent_item_colon" => __( "Parent Yoast Simple Post:", "yoast_e2e_custom_plugin" ),
	];

	$args = [
		"label" => __( "Yoast Simple Posts", "yoast_e2e_custom_plugin" ),
		"labels" => $labels,
		"description" => "",
		"public" => true,
		"publicly_queryable" => true,
		"show_ui" => true,
		"show_in_rest" => true,
		"rest_base" => "",
		"rest_controller_class" => "WP_REST_Posts_Controller",
		"has_archive" => false,
		"show_in_menu" => true,
		"show_in_nav_menus" => true,
		"delete_with_user" => false,
		"exclude_from_search" => false,
		"capability_type" => "post",
		"map_meta_cap" => true,
		"hierarchical" => false,
		"rewrite" => [ "slug" => "yoast_post_type", "with_front" => true ],
		"query_var" => true,
		"supports" => [ "title", "editor", "thumbnail" ],
		"show_in_graphql" => false,
	];

	register_post_type( "yoast_post_type", $args );
}

add_action( 'init', 'yoast_e2e_register_yoast_post_type' );

function yoast_e2e_register_yoast_post_type_taxonomy() {

	/**
	 * Taxonomy: YSP Taxonomies.
	 */

	$labels = [
		"name" => __( "YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"singular_name" => __( "YSP Taxonomy", "yoast_e2e_custom_plugin" ),
		"menu_name" => __( "YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"all_items" => __( "All YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"edit_item" => __( "Edit YSP Taxonomy", "yoast_e2e_custom_plugin" ),
		"view_item" => __( "View YSP Taxonomy", "yoast_e2e_custom_plugin" ),
		"update_item" => __( "Update YSP Taxonomy name", "yoast_e2e_custom_plugin" ),
		"add_new_item" => __( "Add new YSP Taxonomy", "yoast_e2e_custom_plugin" ),
		"new_item_name" => __( "New YSP Taxonomy name", "yoast_e2e_custom_plugin" ),
		"parent_item" => __( "Parent YSP Taxonomy", "yoast_e2e_custom_plugin" ),
		"parent_item_colon" => __( "Parent YSP Taxonomy:", "yoast_e2e_custom_plugin" ),
		"search_items" => __( "Search YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"popular_items" => __( "Popular YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"separate_items_with_commas" => __( "Separate YSP Taxonomies with commas", "yoast_e2e_custom_plugin" ),
		"add_or_remove_items" => __( "Add or remove YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"choose_from_most_used" => __( "Choose from the most used YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"not_found" => __( "No YSP Taxonomies found", "yoast_e2e_custom_plugin" ),
		"no_terms" => __( "No YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"items_list_navigation" => __( "YSP Taxonomies list navigation", "yoast_e2e_custom_plugin" ),
		"items_list" => __( "YSP Taxonomies list", "yoast_e2e_custom_plugin" ),
		"back_to_items" => __( "Back to YSP Taxonomies", "yoast_e2e_custom_plugin" ),
	];

	
	$args = [
		"label" => __( "YSP Taxonomies", "yoast_e2e_custom_plugin" ),
		"labels" => $labels,
		"public" => true,
		"publicly_queryable" => true,
		"hierarchical" => false,
		"show_ui" => true,
		"show_in_menu" => true,
		"show_in_nav_menus" => true,
		"query_var" => true,
		"rewrite" => [ 'slug' => 'yoast_simple_posts_taxonomy', 'with_front' => true, ],
		"show_admin_column" => false,
		"show_in_rest" => true,
		"rest_base" => "yoast_simple_posts_taxonomy",
		"rest_controller_class" => "WP_REST_Terms_Controller",
		"show_in_quick_edit" => false,
		"show_in_graphql" => false,
	];
	register_taxonomy( "yoast_simple_posts_taxonomy", [ "yoast_post_type" ], $args );
}
add_action( 'init', 'yoast_e2e_register_yoast_post_type_taxonomy' );
