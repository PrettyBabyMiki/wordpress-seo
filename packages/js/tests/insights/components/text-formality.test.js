import { useSelect } from "@wordpress/data";
import TextFormality from "../../../src/insights/components/text-formality";
import React from "react";
import renderer from "react-test-renderer";

jest.mock( "@wordpress/data", () => (
	{
		useSelect: jest.fn(),
	}
) );

window.wpseoAdminL10n = {
	"shortlinks-insights-upsell-sidebar-text_formality": "https://yoa.st/formality-upsell-sidebar",
	"shortlinks-insights-upsell-metabox-text_formality": "https://yoa.st/formality-upsell-metabox",
	"shortlinks-insights-upsell-elementor-text_formality": "https://yoa.st/formality-upsell-elementor",
	"shortlinks-insights-text_formality_info_free": "https://yoa.st/formality-sidebar-free",
	"shortlinks-insights-text_formality_info_premium": "https://yoa.st/formality-sidebar",
};

/**
 * Mocks the WordPress `useSelect` hook.
 *
 * @param {object} shouldUpsell     Whether the upsell whould be shown or not.
 * @param {string} formalityLevel   The formality level of the text.
 * @param{object} textLength        The object that contains information on the text length.
 *
 * @returns {void}
 */
function mockSelect( shouldUpsell, formalityLevel, textLength ) {
	const select = jest.fn(
		() => (
			{
				getPreference: jest.fn( () => shouldUpsell ),
				getTextFormalityLevel: jest.fn( () => formalityLevel ),
				getTextLength: jest.fn( () => textLength ),
			}
		)
	);

	useSelect.mockImplementation(
		selectFunction => selectFunction( select )
	);
}

describe( "a test for TextFormality component", () => {
	it( "renders the component in sidebar in Free ", () => {
		mockSelect( true, "formal", { count: 0 } );
		const render = renderer.create( <TextFormality location={ "sidebar" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in metabox in Free ", () => {
		mockSelect( true, "formal", { count: 0 } );
		const render = renderer.create( <TextFormality location={ "metabox" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in sidebar in Premium ", () => {
		mockSelect( false, "formal", { count: 100 } );
		const render = renderer.create( <TextFormality location={ "sidebar" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in sidebar in Premium ", () => {
		mockSelect( false, "informal", { count: 0 } );
		const render = renderer.create( <TextFormality location={ "sidebar" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in metabox in Premium ", () => {
		mockSelect( false, "formal", { count: 0 } );
		const render = renderer.create( <TextFormality location={ "metabox" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in elementor in Free ", () => {
		mockSelect( true, "formal", { count: 0 } );
		const render = renderer.create( <TextFormality location={ "elementor" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "renders the component in elementor in Premium ", () => {
		mockSelect( false, "informal", { count: 1000 } );
		const render = renderer.create( <TextFormality location={ "elementor" } /> );

		const tree = render.toJSON();

		expect( tree ).toMatchSnapshot();
	} );
} );
