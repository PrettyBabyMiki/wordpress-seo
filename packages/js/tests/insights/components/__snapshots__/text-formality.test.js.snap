// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a test for TextFormality component renders the component in elementor in Free  1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      <b>
        Yoast SEO Premium
      </b>
       will help you assess the formality level of your text.
    </p>
    <a
      className="yoast-button yoast-button-upsell"
      href="https://yoa.st/formality-upsell-elementor"
      rel={null}
      target="_blank"
    >
      Unlock with Premium
      <span
        aria-hidden="true"
        className="yoast-button-upsell__caret"
      />
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </div>
  <p>
    <a
      href="https://yoa.st/formality-free"
      rel={null}
      target="_blank"
    >
      Read more about text formality.
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in elementor in Premium  1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

.c1 {
  color: #a4286a;
  font-weight: 600;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      Overall, your text appears to be 
      <span
        className="c1"
      >
        informal
      </span>
      .
    </p>
  </div>
  <p>
    <a
      href="https://yoa.st/formality"
      rel={null}
      target="_blank"
    >
      Read our article on text formality to learn more about how to change the formality level of a text.
      <span
        className="c2"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in metabox in Free  1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      <b>
        Yoast SEO Premium
      </b>
       will help you assess the formality level of your text.
    </p>
    <a
      className="yoast-button yoast-button-upsell"
      href="https://yoa.st/formality-upsell-metabox"
      rel={null}
      target="_blank"
    >
      Unlock with Premium
      <span
        aria-hidden="true"
        className="yoast-button-upsell__caret"
      />
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </div>
  <p>
    <a
      href="https://yoa.st/formality-free"
      rel={null}
      target="_blank"
    >
      Read more about text formality.
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in metabox in Premium  1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      Once you add a bit more copy, we'll be able to tell you the formality level of your text.
    </p>
  </div>
  <p>
    <a
      href="https://yoa.st/formality"
      rel={null}
      target="_blank"
    >
      Read our article on text formality to learn more about how to change the formality level of a text.
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in sidebar in Free  1`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      <b>
        Yoast SEO Premium
      </b>
       will help you assess the formality level of your text.
    </p>
    <a
      className="yoast-button yoast-button-upsell"
      href="https://yoa.st/formality-upsell-sidebar"
      rel={null}
      target="_blank"
    >
      Unlock with Premium
      <span
        aria-hidden="true"
        className="yoast-button-upsell__caret"
      />
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </div>
  <p>
    <a
      href="https://yoa.st/formality-free"
      rel={null}
      target="_blank"
    >
      Read more about text formality.
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in sidebar in Premium  1`] = `
.c2 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

.c1 {
  color: #a4286a;
  font-weight: 600;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      Overall, your text appears to be 
      <span
        className="c1"
      >
        formal
      </span>
      .
    </p>
  </div>
  <p>
    <a
      href="https://yoa.st/formality"
      rel={null}
      target="_blank"
    >
      Read our article on text formality to learn more about how to change the formality level of a text.
      <span
        className="c2"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;

exports[`a test for TextFormality component renders the component in sidebar in Premium  2`] = `
.c1 {
  border: 0;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute !important;
  width: 1px;
  word-wrap: normal !important;
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  margin: 0 2px 0 4px;
}

<div
  className="yoast-text-formality"
>
  <div
    className="yoast-field-group__title"
  >
    <b>
      Text formality
    </b>
    <span
      className="yoast-badge c0 yoast-beta-badge"
    >
      Beta
    </span>
  </div>
  <div>
    <p>
      Once you add a bit more copy, we'll be able to tell you the formality level of your text.
    </p>
  </div>
  <p>
    <a
      href="https://yoa.st/formality"
      rel={null}
      target="_blank"
    >
      Read our article on text formality to learn more about how to change the formality level of a text.
      <span
        className="c1"
      >
        (Opens in a new browser tab)
      </span>
    </a>
  </p>
</div>
`;
